### ISSUE TITLE
Compilation error with 4.13 kernel: "duplicate struct iovec"

### ISSUE THREAD

**Issue by gonzoleeman** [open-iscsi/tcmu-runner#303]
I noticed in our factory build, which just moved to a 4.13 kernel, that the build of tcmu-runner was failing with an error (from the bug report):

```
> [   88s] /usr/bin/cc -Dtcmu_static_EXPORTS -I/usr/include/libnl3 -I/usr/include/gio-unix-2.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include  -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -g -DNDEBUG -Wall -std=c99 -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/tcmu_static.dir/libtcmu-register.c.o   -c /home/abuild/rpmbuild/BUILD/tcmu-runner-1.2.0/libtcmu-register.c
> [   88s] In file included from /usr/include/linux/target_core_user.h:7:0,
> [   88s]                  from /home/abuild/rpmbuild/BUILD/tcmu-runner-1.2.0/libtcmu.c:31:
> [   88s] /usr/include/linux/uio.h:16:8: error: redefinition of 'struct iovec'
> [   88s]  struct iovec
> [   88s]         ^~~~~
> [   88s] In file included from /usr/include/bits/fcntl-linux.h:38:0,
> [   88s]                  from /usr/include/bits/fcntl.h:61,
> [   88s]                  from /usr/include/fcntl.h:35,
> [   88s]                  from /home/abuild/rpmbuild/BUILD/tcmu-runner-1.2.0/libtcmu.c:20:
> [   88s] /usr/include/bits/types/struct_iovec.h:26:8: note: originally defined here
> [   88s]  struct iovec
> [   88s]         ^~~~~
> [   88s] make[2]: *** [CMakeFiles/tcmu_static.dir/build.make:97: CMakeFiles/tcmu_static.dir/libtcmu.c.o] Error 1
```
See [openSUSE Bug 1053501](https://bugzilla.suse.com/show_bug.cgi?id=1053501) for reference.

I'm not sure this is an issue with the latest tcmu-runner as openSUSE is using version 1.2.0 with patches.

I plan to test this with the latest openSUSE Factory 4.13 kernel and the latest tcmu-runner, but I suspect the same issue will occur.

See upstream patch [[PATCH 2/2 uapi: add a compatibility layer between linux/uio.h and glibc]](https://www.spinics.net/lists/netdev/msg422126.html), which I believe will be needed to fix this issue, as well as one change to target_core_user_local.h.

Has anybody else seen this issue?

**Comment by gonzoleeman** [comment]
**NOTE** I _believe_ I have a solution for this. Well, at least it allows tcmu-runner to compile correctly. I will post a pull request so that you can look at the proposed changes. It boils down to moving some includes around (in order), and including <sys/uio.h> instead of <linux/uio.h> in target_core_user.h. This allows tcmu-runner to compile even though the upstream issue has not been fixed.

**Comment by gonzoleeman** [comment]
I have created an updated pull request which I believe fixes this issue and will not cause a regression on RedHat, but please test it and let me know.

**Comment by denis-koptev** [comment]
Yes, replacing `<linux/uio.h>` with `<sys/uio.h>` in `target_core_user_local.h` helps to build it successfully. Thanks.

**Comment by denis-koptev** [comment]
I suppose that the latest kernel versions now have `iovec` definition in `<sys/uio.h>`. (Ubuntu 17.10, 18.04 confirmed). On the other hand, old kernel doesn't have this definition, so `<linux/uio.h>` works fine.
